File: src/types/pcb.ts
// ═══════════════════════════════════════════════════════════════
// PCB Data Model — KiCad .kicad_pcb compatible
// ═══════════════════════════════════════════════════════════════

export interface PcbPoint {
  x: number;
  y: number;
}

// --- Layer System ---

export type CopperLayerId =
  | "F.Cu" | "In1.Cu" | "In2.Cu" | "In3.Cu" | "In4.Cu" | "In5.Cu"
  | "In6.Cu" | "In7.Cu" | "In8.Cu" | "In9.Cu" | "In10.Cu" | "In11.Cu"
  | "In12.Cu" | "In13.Cu" | "In14.Cu" | "In15.Cu" | "In16.Cu" | "In17.Cu"
  | "In18.Cu" | "In19.Cu" | "In20.Cu" | "In21.Cu" | "In22.Cu" | "In23.Cu"
  | "In24.Cu" | "In25.Cu" | "In26.Cu" | "In27.Cu" | "In28.Cu" | "In29.Cu"
  | "In30.Cu" | "B.Cu";

export type TechLayerId =
  | "F.SilkS" | "B.SilkS"
  | "F.Mask" | "B.Mask"
  | "F.Paste" | "B.Paste"
  | "F.Fab" | "B.Fab"
  | "F.CrtYd" | "B.CrtYd"
  | "F.Adhes" | "B.Adhes"
  | "Edge.Cuts"
  | "Margin"
  | "Dwgs.User" | "Cmts.User" | "Eco1.User" | "Eco2.User";

export type PcbLayerId = CopperLayerId | TechLayerId | string;

export interface PcbLayer {
  id: PcbLayerId;
  name: string;
  type: "signal" | "power" | "mixed" | "jumper" | "user";
  visible: boolean;
  color: string;
  opacity: number;
}

export interface LayerStack {
  layers: PcbLayer[];
  copperCount: number; // 2, 4, 6, ... 32
}

// --- Board ---

export interface PcbBoard {
  uuid: string;
  version: string;
  generator: string;
  thickness: number; // Board thickness in mm
  outline: PcbPoint[]; // Board outline polygon
  layers: LayerStack;
  setup: PcbSetup;
}

export interface PcbSetup {
  gridSize: number;
  traceWidth: number; // Default trace width
  viaDiameter: number;
  viaDrill: number;
  clearance: number; // Default clearance
  trackMinWidth: number;
  viaMinDiameter: number;
  viaMinDrill: number;
  copperFinish: string;
}

// --- Footprints (Components on PCB) ---

export interface PcbFootprint {
  uuid: string;
  reference: string; // "U1", "R1", etc.
  value: string;
  footprintId: string; // Library footprint ID
  position: PcbPoint;
  rotation: number;
  layer: PcbLayerId; // "F.Cu" or "B.Cu"
  locked: boolean;
  pads: PcbPad[];
  graphics: PcbGraphic[];
  courtyard?: PcbPoint[]; // Courtyard polygon
  model3d?: string; // STEP model path
}

export interface PcbPad {
  uuid: string;
  number: string; // "1", "2", "A1", etc.
  type: "thru_hole" | "smd" | "np_thru_hole" | "connect";
  shape: "circle" | "rect" | "oval" | "roundrect" | "custom" | "trapezoid";
  position: PcbPoint; // Relative to footprint
  size: [number, number]; // [width, height]
  drill?: { diameter: number; shape?: "circle" | "oval"; offset?: PcbPoint };
  layers: PcbLayerId[];
  net?: { number: number; name: string };
  roundrectRatio?: number;
  thermalGap?: number;
  thermalWidth?: number;
  solderMaskExpansion?: number;
  pasteMaskExpansion?: number;
}

export type PcbGraphic =
  | { type: "line"; start: PcbPoint; end: PcbPoint; layer: PcbLayerId; width: number }
  | { type: "rect"; start: PcbPoint; end: PcbPoint; layer: PcbLayerId; width: number; fill?: boolean }
  | { type: "circle"; center: PcbPoint; radius: number; layer: PcbLayerId; width: number; fill?: boolean }
  | { type: "arc"; start: PcbPoint; mid: PcbPoint; end: PcbPoint; layer: PcbLayerId; width: number }
  | { type: "poly"; points: PcbPoint[]; layer: PcbLayerId; width: number; fill?: boolean }
  | { type: "text"; text: string; position: PcbPoint; layer: PcbLayerId; fontSize: number; rotation: number; mirror?: boolean };

// --- Traces & Vias ---

export interface PcbTrace {
  uuid: string;
  net: number; // Net number
  segments: PcbSegment[];
}

export interface PcbSegment {
  uuid: string;
  start: PcbPoint;
  end: PcbPoint;
  width: number;
  layer: PcbLayerId;
  net: number;
}

export interface PcbVia {
  uuid: string;
  position: PcbPoint;
  diameter: number;
  drill: number;
  layers: [PcbLayerId, PcbLayerId]; // [start_layer, end_layer]
  net: number;
  type: "through" | "blind" | "buried" | "micro";
}

// --- Copper Zones (Pours) ---

export interface PcbZone {
  uuid: string;
  net: number;
  netName: string;
  layer: PcbLayerId;
  outline: PcbPoint[];
  priority: number;
  fillType: "solid" | "hatch" | "none";
  thermalRelief: boolean;
  thermalGap: number;
  thermalWidth: number;
  clearance: number;
  minThickness: number;
  keepout?: {
    tracks: boolean;
    vias: boolean;
    pads: boolean;
    copperpour: boolean;
    footprints: boolean;
  };
  filled?: PcbPoint[][]; // Filled polygons (computed)
}

// --- Nets ---

export interface PcbNet {
  number: number;
  name: string;
}

// --- Ratsnest ---

export interface RatsnestLine {
  padA: { footprintUuid: string; padNumber: string };
  padB: { footprintUuid: string; padNumber: string };
  net: number;
}

// --- Design Rules ---

export interface PcbDesignRule {
  uuid: string;
  name: string;
  type: "clearance" | "trace_width" | "via_size" | "diff_pair" | "annular_ring" | "hole_size";
  scope: {
    kind: "all" | "net_class" | "net" | "layer";
    target?: string;
  };
  min?: number;
  preferred?: number;
  max?: number;
  enabled: boolean;
}

// --- DRC ---

export interface DrcViolation {
  uuid: string;
  type: "clearance" | "unrouted" | "short_circuit" | "annular_ring" | "minimum_width" | "drill_size" | "board_outline";
  severity: "error" | "warning";
  message: string;
  position: PcbPoint;
  layer?: PcbLayerId;
  objects: string[]; // UUIDs of involved objects
}

// --- Full PCB Document ---

export interface PcbData {
  board: PcbBoard;
  footprints: PcbFootprint[];
  segments: PcbSegment[]; // All trace segments (flat, not grouped by trace)
  vias: PcbVia[];
  zones: PcbZone[];
  nets: PcbNet[];
  graphics: PcbGraphic[]; // Board-level graphics (dimensions, drawings)
  texts: { uuid: string; text: string; position: PcbPoint; layer: PcbLayerId; fontSize: number; rotation: number }[];
  designRules: PcbDesignRule[];
}

// --- Layer Colors (Altium-inspired) ---

export const DEFAULT_LAYER_COLORS: Record<string, string> = {
  "F.Cu": "#cc0000",     // Red
  "B.Cu": "#0000cc",     // Blue
  "In1.Cu": "#c8c800",   // Yellow
  "In2.Cu": "#00c8c8",   // Cyan
  "In3.Cu": "#c800c8",   // Magenta
  "In4.Cu": "#00c800",   // Green
  "F.SilkS": "#c8c8c8",  // Light gray
  "B.SilkS": "#808080",  // Dark gray
  "F.Mask": "#800080",   // Purple (transparent)
  "B.Mask": "#008080",   // Teal (transparent)
  "F.Paste": "#808000",  // Olive
  "B.Paste": "#008000",  // Dark green
  "F.Fab": "#aaaa00",    // Yellow-green
  "B.Fab": "#00aaaa",    // Cyan-green
  "F.CrtYd": "#c8c800",  // Yellow outline
  "B.CrtYd": "#00c8c8",  // Cyan outline
  "Edge.Cuts": "#c8c800", // Yellow
  "Dwgs.User": "#808080",
  "Cmts.User": "#606060",
};

File: crates/signex-types/src/pcb.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::property::PcbProperty;

pub use crate::schematic::Point;

// ---------------------------------------------------------------------------
// Pad / Via enums
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PadType {
    Thru,
    Smd,
    Connect,
    NpThru,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PadShape {
    Circle,
    Rect,
    Oval,
    Trapezoid,
    RoundRect,
    Custom,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ViaType {
    Through,
    Blind,
    Micro,
}

// ---------------------------------------------------------------------------
// Layer definition
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LayerDef {
    pub id: u8,
    pub name: String,
    pub layer_type: String,
}

// ---------------------------------------------------------------------------
// PCB setup
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PcbSetup {
    #[serde(default)]
    pub grid_size: f64,
    #[serde(default)]
    pub trace_width: f64,
    #[serde(default)]
    pub via_diameter: f64,
    #[serde(default)]
    pub via_drill: f64,
    #[serde(default)]
    pub clearance: f64,
    #[serde(default)]
    pub track_min_width: f64,
    #[serde(default)]
    pub via_min_diameter: f64,
    #[serde(default)]
    pub via_min_drill: f64,
}

// ---------------------------------------------------------------------------
// Net definition
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetDef {
    pub number: u32,
    pub name: String,
}

// ---------------------------------------------------------------------------
// Drill definition
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DrillDef {
    pub diameter: f64,
    #[serde(default)]
    pub shape: String,
}

// ---------------------------------------------------------------------------
// Pad net reference
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PadNet {
    pub number: u32,
    pub name: String,
}

// ---------------------------------------------------------------------------
// Pad
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pad {
    pub uuid: Uuid,
    #[serde(default)]
    pub number: String,
    pub pad_type: PadType,
    pub shape: PadShape,
    pub position: Point,
    pub size: Point,
    pub drill: Option<DrillDef>,
    #[serde(default)]
    pub layers: Vec<String>,
    pub net: Option<PadNet>,
    #[serde(default)]
    pub roundrect_ratio: f64,
}

// ---------------------------------------------------------------------------
// Footprint graphic
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FpGraphic {
    #[serde(default)]
    pub graphic_type: String,
    #[serde(default)]
    pub layer: String,
    #[serde(default)]
    pub width: f64,
    pub start: Option<Point>,
    pub end: Option<Point>,
    pub center: Option<Point>,
    pub mid: Option<Point>,
    #[serde(default)]
    pub radius: f64,
    #[serde(default)]
    pub points: Vec<Point>,
    #[serde(default)]
    pub text: String,
    #[serde(default)]
    pub font_size: f64,
    pub position: Option<Point>,
    #[serde(default)]
    pub rotation: f64,
    #[serde(default)]
    pub fill: String,
}

// ---------------------------------------------------------------------------
// Footprint
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Footprint {
    pub uuid: Uuid,
    #[serde(default)]
    pub reference: String,
    #[serde(default)]
    pub value: String,
    #[serde(default)]
    pub footprint_id: String,
    pub position: Point,
    #[serde(default)]
    pub rotation: f64,
    #[serde(default)]
    pub layer: String,
    #[serde(default)]
    pub locked: bool,
    #[serde(default)]
    pub pads: Vec<Pad>,
    #[serde(default)]
    pub graphics: Vec<FpGraphic>,
    #[serde(default)]
    pub properties: Vec<PcbProperty>,
}

// ---------------------------------------------------------------------------
// Traces / routing
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Segment {
    pub uuid: Uuid,
    pub start: Point,
    pub end: Point,
    #[serde(default)]
    pub width: f64,
    #[serde(default)]
    pub layer: String,
    #[serde(default)]
    pub net: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Via {
    pub uuid: Uuid,
    pub position: Point,
    #[serde(default)]
    pub diameter: f64,
    #[serde(default)]
    pub drill: f64,
    #[serde(default)]
    pub layers: Vec<String>,
    #[serde(default)]
    pub net: u32,
    #[serde(default = "default_via_type")]
    pub via_type: ViaType,
}

fn default_via_type() -> ViaType {
    ViaType::Through
}

// ---------------------------------------------------------------------------
// Zone
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Zone {
    pub uuid: Uuid,
    #[serde(default)]
    pub net: u32,
    #[serde(default)]
    pub net_name: String,
    #[serde(default)]
    pub layer: String,
    #[serde(default)]
    pub outline: Vec<Point>,
    #[serde(default)]
    pub priority: u32,
    #[serde(default)]
    pub fill_type: String,
    #[serde(default)]
    pub thermal_relief: bool,
    #[serde(default)]
    pub thermal_gap: f64,
    #[serde(default)]
    pub thermal_width: f64,
    #[serde(default)]
    pub clearance: f64,
    #[serde(default)]
    pub min_thickness: f64,
}

// ---------------------------------------------------------------------------
// Board-level graphics and text
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BoardGraphic {
    #[serde(default)]
    pub graphic_type: String,
    #[serde(default)]
    pub layer: String,
    #[serde(default)]
    pub width: f64,
    pub start: Option<Point>,
    pub end: Option<Point>,
    pub center: Option<Point>,
    #[serde(default)]
    pub radius: f64,
    #[serde(default)]
    pub points: Vec<Point>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BoardText {
    pub uuid: Uuid,
    #[serde(default)]
    pub text: String,
    pub position: Point,
    #[serde(default)]
    pub layer: String,
    #[serde(default)]
    pub font_size: f64,
    #[serde(default)]
    pub rotation: f64,
}

// ---------------------------------------------------------------------------
// Top-level board
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PcbBoard {
    pub uuid: Uuid,
    #[serde(default)]
    pub version: u32,
    #[serde(default)]
    pub generator: String,
    #[serde(default)]
    pub thickness: f64,
    #[serde(default)]
    pub outline: Vec<Point>,
    #[serde(default)]
    pub layers: Vec<LayerDef>,
    pub setup: Option<PcbSetup>,
    #[serde(default)]
    pub nets: Vec<NetDef>,
    #[serde(default)]
    pub footprints: Vec<Footprint>,
    #[serde(default)]
    pub segments: Vec<Segment>,
    #[serde(default)]
    pub vias: Vec<Via>,
    #[serde(default)]
    pub zones: Vec<Zone>,
    #[serde(default)]
    pub graphics: Vec<BoardGraphic>,
    #[serde(default)]
    pub texts: Vec<BoardText>,
}
